### PR TITLE
Super Admin Details stored.

### PR DESCRIPTION
**Pull request by JitendraSaini1995**
super admin encrypted data(username and password) stored in shared preferences.

### CODE REVIEW

**Comment by JitendraSaini1995** [comment]
[Uploading OT3.webm…]()

**Comment by JitendraSaini1995** [comment]
[Uploading OT3.webm…]()

**Comment by JitendraSaini1995** [comment]
Screenshot 2024-02-20 144623

**Comment by JitendraSaini1995** [comment]
[OT3.webm](https://github.com/sublimitysoftwares/OnTime_Android/assets/131166947/3950aeec-e07d-4797-89d5-70b83ce33ff1)

File: app/src/main/AndroidManifest.xml [hunk 1]
```diff
@@ -1,10 +1,20 @@
...
-    xmlns:tools="http://schemas.android.com/tools">
+    package="com.allocate.ontime"
+    xmlns:tools="http://schemas.android.com/tools"
+    tools:ignore="CoarseFineLocation">
     <uses-permission android:name="android.permission.INTERNET"/>
+    <uses-permission android:name="android.permission.ACCESS_FINE_LOCATION" />
+    <uses-permission android:name = "android.permission.ACCESS_WIFI_STATE" />
+    <uses-permission android:name="android.permission.LOCAL_MAC_ADDRESS"
+        tools:ignore="ProtectedPermissions" />
+    <uses-permission android:name="android.permission.READ_PRIVILEGED_PHONE_STATE"
```
Comment (mohit2988) [inline app/src/main/AndroidManifest.xml:11]: Why this permission required for?

File: app/src/main/java/com/allocate/ontime/business_logic/repository/DeviceInfoRepository.kt [hunk 2]
```diff
@@ -0,0 +1,56 @@
...
+
+class DeviceInfoRepository @Inject constructor(
+    private val deviceInfoApi: DeviceInfoApi,
+    private val superAdminApi: SuperAdminApi
+) {
+    suspend fun getDeviceInfo(): DataOrException<DeviceInfo, Exception> {
+        val dataOrException = DataOrException<DeviceInfo, Exception>()
+        try {
+            dataOrException.data = deviceInfoApi.getDeviceInfo(imei = Utils.imei)
+//            dataOrException.data = deviceInfoApi.getDeviceInfo(imei = "867291070025769")
```
[the remark this answers is not in the document]
Comment (JitendraSaini1995) [reply app/src/main/java/com/allocate/ontime/business_logic/repository/DeviceInfoRepository.kt:24]: done!

File: app/src/main/java/com/allocate/ontime/encryption/EDHelper.java [hunk 3]
```diff
@@ -0,0 +1,76 @@
...
+import com.google.gson.Gson;
+
+import retrofit2.Response;
+//import timber.log.Timber;
+
+/**
+ * Created by admin on 02-Feb-18.
+ */
+
+public class EDHelper {
```
[the remark this answers is not in the document]
Comment (JitendraSaini1995) [reply app/src/main/java/com/allocate/ontime/encryption/EDHelper.java:15]: done!

File: app/src/main/java/com/allocate/ontime/encryption/AESHelper.java [hunk 4]
```diff
@@ -0,0 +1,87 @@
...
+import javax.crypto.Cipher;
+import javax.crypto.IllegalBlockSizeException;
+import javax.crypto.NoSuchPaddingException;
+import javax.crypto.spec.IvParameterSpec;
+import javax.crypto.spec.SecretKeySpec;
+
+/**
+ * Created by Azlaan Khan on 9/25/2015.
+ */
+public class AESHelper {
```
[the remark this answers is not in the document]
Comment (JitendraSaini1995) [reply app/src/main/java/com/allocate/ontime/encryption/AESHelper.java:22]: done!

File: app/src/main/java/com/allocate/ontime/business_logic/utils/Constants.kt [hunk 5]
```diff
@@ -15,4 +18,9 @@ object Constants {
       const val AdminRegistrationScreen = "AdminRegistrationScreen"
       const val FobRegisterScreen = "FobRegisterScreen"
       const val SuperAdminSettingScreen = "SuperAdminSettingScreen"
+
+      const val UserName = "UserName"
```
[the remark this answers is not in the document]
Comment (JitendraSaini1995) [reply app/src/main/java/com/allocate/ontime/business_logic/utils/Constants.kt:22]: done!

File: app/src/main/java/com/allocate/ontime/business_logic/utils/Utils.kt [hunk 6]
```diff
@@ -0,0 +1,11 @@
+package com.allocate.ontime.business_logic.utils
+
+
+import javax.inject.Singleton
+
+@Singleton
+object Utils {
```
[the remark this answers is not in the document]
Comment (JitendraSaini1995) [reply app/src/main/java/com/allocate/ontime/business_logic/utils/Utils.kt:7]: done!

File: app/src/main/java/com/allocate/ontime/business_logic/viewmodel/splash/SplashViewModel.kt [hunk 7]
```diff
@@ -17,21 +24,24 @@ import kotlinx.coroutines.flow.MutableStateFlow
...
 ) : ViewModel() {
 
-
     private val _acknowledgementStatus = MutableStateFlow(0)
     val acknowledgementStatus = _acknowledgementStatus.asStateFlow()
 
     init {
         viewModelScope.launch(Dispatchers.IO) {
             val deviceInfoApiData = async { repository.getDeviceInfo() }.await()
+            Log.d("data1", "hello:$deviceInfoApiData")
```
[the remark this answers is not in the document]
Comment (JitendraSaini1995) [reply app/src/main/java/com/allocate/ontime/business_logic/viewmodel/splash/SplashViewModel.kt:44]: done!

File: app/src/main/java/com/allocate/ontime/business_logic/viewmodel/MainViewModel.kt [hunk 8]
```diff
@@ -0,0 +1,40 @@
...
+class MainViewModel @Inject constructor(
+    private val repository: DeviceInfoRepository,
+    context: Context
+) : ViewModel() {
+
+    init {
+        viewModelScope.launch {
+            val result = repository.postSuperAdminDetails()
+            if (result.data?.isSuccessful == true) {
+                val data = result.data!!.body()?.data
```
[the remark this answers is not in the document]
Comment (JitendraSaini1995) [reply app/src/main/java/com/allocate/ontime/business_logic/viewmodel/MainViewModel.kt:26]: done!

File: app/src/main/AndroidManifest.xml (hunk 1 above)
Comment (JitendraSaini1995) [reply app/src/main/AndroidManifest.xml:11]: If i am not use this permission then it gives compile time errors.

File: app/src/main/java/com/allocate/ontime/business_logic/di/AppModule.kt [hunk 9]
```diff
@@ -39,19 +64,35 @@ object AppModule {
             .fallbackToDestructiveMigration()
             .build()
 
-    // It provides the dependency of OnTimeApi Interface.
+    //     It provides the dependency of OnTimeApi Interface.
     @Singleton
```
[the remark this answers is not in the document]
Comment (JitendraSaini1995) [reply app/src/main/java/com/allocate/ontime/business_logic/di/AppModule.kt:68]: https://developer.android.com/training/dependency-injection/hilt-android#component-scopes

File: app/src/main/java/com/allocate/ontime/business_logic/data/shared_preferences/PreferencesManager.kt [hunk 10]
```diff
@@ -0,0 +1,19 @@
+package com.allocate.ontime.business_logic.data.shared_preferences
+
+import android.content.Context
+import android.content.SharedPreferences
+
+class PreferencesManager(context: Context) {
+    private val sharedPreferences: SharedPreferences =
```
[the remark this answers is not in the document]
Comment (JitendraSaini1995) [reply app/src/main/java/com/allocate/ontime/business_logic/data/shared_preferences/PreferencesManager.kt:7]: done!

File: app/src/main/java/com/allocate/ontime/business_logic/utils/Constants.kt [hunk 11]
```diff
@@ -1,12 +1,15 @@
 package com.allocate.ontime.business_logic.utils
 
+
 object Constants {
-      const val BASE_URL = "http://api.cpair.uk/"
+//      const val BASE_URL = "http://api.cpair.uk/"
+        const val BASE_URL = "http://cpaapi.sublimitysoft.com/"
```
[the remark this answers is not in the document]
Comment (JitendraSaini1995) [reply app/src/main/java/com/allocate/ontime/business_logic/utils/Constants.kt:6]: done!

File: app/src/main/java/com/allocate/ontime/MainActivity.kt [hunk 12]
```diff
@@ -38,17 +43,38 @@ class MainActivity : ComponentActivity() {
...
 
 
-    @SuppressLint("CoroutineCreationDuringComposition")
     override fun onCreate(savedInstanceState: Bundle?) {
         super.onCreate(savedInstanceState)
         setContent {
             OnTimeApp()
         }
+
+        if (ActivityCompat.checkSelfPermission(
```
[the remark this answers is not in the document]
Comment (JitendraSaini1995) [reply app/src/main/java/com/allocate/ontime/MainActivity.kt:61]: I removed this code.I have got IMEI without checking this permission.

File: app/src/main/java/com/allocate/ontime/MainActivity.kt [hunk 13]
```diff
@@ -69,6 +95,20 @@ class MainActivity : ComponentActivity() {
...
+
+        // Check if the device is running Android 10 or higher
+        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.Q) {
+            // Use getImei() for Android 10 and higher
+            imei = telephonyManager.imei
+        } else {
+            // Use getDeviceId() for Android 9 and below
+            imei = telephonyManager.deviceId
+        }
+        Utils.imei = imei
```
[the remark this answers is not in the document]
Comment (JitendraSaini1995) [reply app/src/main/java/com/allocate/ontime/MainActivity.kt:110]: done!

File: app/src/main/java/com/allocate/ontime/business_logic/utils/Utils.kt [hunk 14]
```diff
@@ -0,0 +1,11 @@
...
+    var imei: String? = null
```
[the remark this answers is not in the document]
Comment (JitendraSaini1995) [reply app/src/main/java/com/allocate/ontime/business_logic/utils/Utils.kt:8]: done!

File: app/src/main/java/com/allocate/ontime/business_logic/viewmodel/splash/SplashViewModel.kt [hunk 15]
```diff
@@ -49,22 +59,26 @@ class SplashViewModel @Inject constructor(
...
                         locationCode = data.LocationCode,
                         acknowledgementStatus = 0,
                         sites = data.Sites,
                         siteName = data.SiteName,
                         isRLD = data.IsRLD
+
                     )
                 )
+                val asApiURL: String = data.ASApiURL
+                Utils.asApiURL = asApiURL
```
[the remark this answers is not in the document]
Comment (JitendraSaini1995) [reply app/src/main/java/com/allocate/ontime/business_logic/viewmodel/splash/SplashViewModel.kt:72]: done!

File: app/src/main/java/com/allocate/ontime/business_logic/viewmodel/splash/SplashViewModel.kt [hunk 16]
```diff
@@ -124,6 +138,13 @@ class SplashViewModel @Inject constructor(
         }
     }
 
+    @SuppressLint("HardwareIds")
+    private fun getMacAddress(context: Context): String {
```
[the remark this answers is not in the document]
Comment (JitendraSaini1995) [reply app/src/main/java/com/allocate/ontime/business_logic/viewmodel/splash/SplashViewModel.kt:142]: done!

File: app/src/main/java/com/allocate/ontime/business_logic/viewmodel/splash/SplashViewModel.kt [hunk 17]
```diff
@@ -49,22 +59,26 @@ class SplashViewModel @Inject constructor(
...
             }
-            observeAllDeviceInfo(this, deviceInfoApiData)
+            observeAllDeviceInfo(this, deviceInfoApiData, context)
         }
     }
 
     private suspend fun observeAllDeviceInfo(
         coroutineScope: CoroutineScope,
         getDeviceInfoApiData: DataOrException<DeviceInfo, Exception>,
+        context: Context
```
[the remark this answers is not in the document]
Comment (JitendraSaini1995) [reply app/src/main/java/com/allocate/ontime/business_logic/viewmodel/splash/SplashViewModel.kt:81]: when i am using context directly.it is giving error.

File: app/src/main/java/com/allocate/ontime/business_logic/di/AppModule.kt [hunk 18]
```diff
@@ -35,23 +69,39 @@ object AppModule {
...
-            .baseUrl(Constants.BASE_URL)
+            .baseUrl(BuildConfig.BASE_URL)
             .addConverterFactory(GsonConverterFactory.create())
             .build()
     }
+
+    @Provides
+    @SuperAdminRetrofit
+    fun provideRetrofitClient2(): Retrofit {
+        return Constants.asApiURL.let {
```
Comment (mohit2988) [inline app/src/main/java/com/allocate/ontime/business_logic/di/AppModule.kt:100]: Constants.asApiURL, Try to do a better way for this.
You can keep it in shared preference if we need to keep it

File: app/src/main/java/com/allocate/ontime/business_logic/repository/DeviceInfoRepository.kt [hunk 19]
```diff
@@ -0,0 +1,56 @@
...
+            Log.d("EXC", "getAllDeviceInfo: ${dataOrException.e}")
+        }
+        return dataOrException
+    }
+
+    suspend fun postSuperAdminDetails(): DataOrException<Response<EDModel>, Exception> {
+        val dataOrException = DataOrException<Response<EDModel>, Exception>()
+        try {
+            dataOrException.data = superAdminApi.getCISuperAdminDetails()
+            Log.d("SuperAdmin", "${dataOrException.data}")
```
Comment (mohit2988) [inline app/src/main/java/com/allocate/ontime/business_logic/repository/DeviceInfoRepository.kt:49]: use TAG here

like 

companion object {
const val TAG = "DeviceInfoRepository"
}

Comment (mohit2988) [reply app/src/main/java/com/allocate/ontime/business_logic/repository/DeviceInfoRepository.kt:49]: do not print "dataOrException.data", you can print here getCISuperAdminDetails success.

File: app/src/main/java/com/allocate/ontime/business_logic/utils/Constants.kt [hunk 20]
```diff
@@ -1,18 +1,22 @@
 package com.allocate.ontime.business_logic.utils
 
 object Constants {
-      const val BASE_URL = "http://api.cpair.uk/"
-      const val databaseName = "OnTime_DB"
-      const val getDeviceInfoUrlEndPoint = "GetDevice?IMEI=864392050058185"
-      const val editDeviceInfoUrlEndPoint = "editDeviceAppData"
-
+      var asApiURL: String = "http://airstack.sublimitysoft.com/AirstackAPI/API/"
```
Comment (mohit2988) [inline app/src/main/java/com/allocate/ontime/business_logic/utils/Constants.kt:4]: remove it

File: app/src/main/java/com/allocate/ontime/business_logic/repository/DeviceInfoRepository.kt [hunk 21]
```diff
@@ -0,0 +1,56 @@
...
+
+    suspend fun postDeviceInfo(
+        appInfo: AppInfo
+    ): DataOrException<EditDeviceInfo, Exception> {
+        val dataOrException = DataOrException<EditDeviceInfo, Exception>()
+        try {
+            dataOrException.data = deviceInfoApi.postEditDeviceInfo(appInfo)
+        } catch (exception: Exception) {
+            dataOrException.e = exception
+            Log.d("EXC", "getAllDeviceInfo: ${dataOrException.e}")
```
Comment (mohit2988) [inline app/src/main/java/com/allocate/ontime/business_logic/repository/DeviceInfoRepository.kt:40]: use Log.i()

File: app/src/main/java/com/allocate/ontime/business_logic/viewmodel/MainViewModel.kt [hunk 22]
```diff
@@ -0,0 +1,39 @@
...
+    context: Context
+) : ViewModel() {
+
+    init {
+        viewModelScope.launch {
+            val result = repository.postSuperAdminDetails()
+            if (result.data?.isSuccessful == true) {
+                val data = result.data?.body()?.data
+                val decryptedData =
+                    EDModel(data.toString()).getResponseModel(SuperAdminResponse::class.java)
```
Comment (mohit2988) [inline app/src/main/java/com/allocate/ontime/business_logic/viewmodel/MainViewModel.kt:27]: if result.data is null then?

File: app/src/main/java/com/allocate/ontime/business_logic/viewmodel/splash/SplashViewModel.kt [hunk 23]
```diff
@@ -17,21 +25,26 @@ import kotlinx.coroutines.flow.MutableStateFlow
...
+    private val repository: DeviceInfoRepository,
     private val daoRepository: DaoRepository,
+    private val deviceUtils: DeviceUtils,
+    context: Context,
 ) : ViewModel() {
 
-
     private val _acknowledgementStatus = MutableStateFlow(0)
     val acknowledgementStatus = _acknowledgementStatus.asStateFlow()
+    private val TAG = "SplashViewModel"
```
Comment (mohit2988) [inline app/src/main/java/com/allocate/ontime/business_logic/viewmodel/splash/SplashViewModel.kt:42]: use companion

File: app/src/main/java/com/allocate/ontime/encryption/AESHelper.kt [hunk 24]
```diff
@@ -0,0 +1,118 @@
...
+import java.security.NoSuchAlgorithmException
+import javax.crypto.BadPaddingException
+import javax.crypto.Cipher
+import javax.crypto.IllegalBlockSizeException
+import javax.crypto.NoSuchPaddingException
+import javax.crypto.spec.IvParameterSpec
+import javax.crypto.spec.SecretKeySpec
+
+/**
+ * Created by Azlaan Khan on 9/25/2015.
```
Comment (mohit2988) [inline app/src/main/java/com/allocate/ontime/encryption/AESHelper.kt:18]: Remove this and Submility and change the date

File: app/src/main/java/com/allocate/ontime/encryption/EDHelper.kt [hunk 25]
```diff
@@ -0,0 +1,68 @@
+package com.allocate.ontime.encryption
+
+import android.util.Log
+import com.google.gson.Gson
+import retrofit2.Response
+
+/**
+ * Created by admin on 02-Feb-18.
```
Comment (mohit2988) [inline app/src/main/java/com/allocate/ontime/encryption/EDHelper.kt:8]: do the same here

File: app/src/main/java/com/allocate/ontime/encryption/EDHelper.kt [hunk 26]
```diff
@@ -0,0 +1,68 @@
...
+ */
+
+object EDHelper {
+    val TAG = "EDHelper"
+    const val KEY = "MDEyMzQ1Njc4OTAxMjM0NTY3ODkwMTIzNDU2Nzg5MzA="
```
Comment (mohit2988) [inline app/src/main/java/com/allocate/ontime/encryption/EDHelper.kt:13]: Put this in gradle.properties and use it here

File: app/src/main/java/com/allocate/ontime/presentation_logic/screens/super_admin/SuperAdminSettingScreen.kt [hunk 27]
```diff
@@ -381,18 +383,21 @@ private fun SuperAdminSettingInfo(
                 Text(
                     text = "Site Name",
```
Comment (mohit2988) [inline app/src/main/java/com/allocate/ontime/presentation_logic/screens/super_admin/SuperAdminSettingScreen.kt:384]: put this in string resource file, find every hard coded UI string and put all in string resource file

File: app/src/main/java/com/allocate/ontime/business_logic/repository/DeviceInfoRepository.kt (hunk 19 above)
Comment (JitendraSaini1995) [reply app/src/main/java/com/allocate/ontime/business_logic/repository/DeviceInfoRepository.kt:49]: done!

File: app/src/main/java/com/allocate/ontime/business_logic/repository/DeviceInfoRepository.kt [hunk 28]
```diff
@@ -0,0 +1,56 @@
...
+    }
+
+    suspend fun postSuperAdminDetails(): DataOrException<Response<EDModel>, Exception> {
+        val dataOrException = DataOrException<Response<EDModel>, Exception>()
+        try {
+            dataOrException.data = superAdminApi.getCISuperAdminDetails()
+            Log.d("SuperAdmin", "${dataOrException.data}")
+        } catch (exception: Exception) {
+            dataOrException.e = exception
+            Log.d("EXC", "getSuperAdminDetails: ${dataOrException.e}")
```
[the remark this answers is not in the document]
Comment (JitendraSaini1995) [reply app/src/main/java/com/allocate/ontime/business_logic/repository/DeviceInfoRepository.kt:52]: done!

File: app/src/main/java/com/allocate/ontime/business_logic/repository/DeviceInfoRepository.kt (hunk 21 above)
Comment (JitendraSaini1995) [reply app/src/main/java/com/allocate/ontime/business_logic/repository/DeviceInfoRepository.kt:40]: done!

File: app/src/main/java/com/allocate/ontime/business_logic/viewmodel/MainViewModel.kt (hunk 22 above)
Comment (JitendraSaini1995) [reply app/src/main/java/com/allocate/ontime/business_logic/viewmodel/MainViewModel.kt:27]: done!

File: app/src/main/java/com/allocate/ontime/business_logic/viewmodel/splash/SplashViewModel.kt (hunk 23 above)
Comment (JitendraSaini1995) [reply app/src/main/java/com/allocate/ontime/business_logic/viewmodel/splash/SplashViewModel.kt:42]: done!

File: app/src/main/java/com/allocate/ontime/encryption/AESHelper.kt (hunk 24 above)
Comment (JitendraSaini1995) [reply app/src/main/java/com/allocate/ontime/encryption/AESHelper.kt:18]: done!

File: app/src/main/java/com/allocate/ontime/encryption/EDHelper.kt (hunk 25 above)
Comment (JitendraSaini1995) [reply app/src/main/java/com/allocate/ontime/encryption/EDHelper.kt:8]: done!

File: app/src/main/java/com/allocate/ontime/presentation_logic/screens/super_admin/SuperAdminSettingScreen.kt (hunk 27 above)
Comment (JitendraSaini1995) [reply app/src/main/java/com/allocate/ontime/presentation_logic/screens/super_admin/SuperAdminSettingScreen.kt:384]: done!

File: app/src/main/java/com/allocate/ontime/encryption/EDHelper.kt (hunk 26 above)
Comment (JitendraSaini1995) [reply app/src/main/java/com/allocate/ontime/encryption/EDHelper.kt:13]: done!

File: app/src/main/java/com/allocate/ontime/business_logic/di/AppModule.kt (hunk 18 above)
Comment (JitendraSaini1995) [reply app/src/main/java/com/allocate/ontime/business_logic/di/AppModule.kt:100]: done!

File: app/src/main/java/com/allocate/ontime/business_logic/utils/Constants.kt (hunk 20 above)
Comment (JitendraSaini1995) [reply app/src/main/java/com/allocate/ontime/business_logic/utils/Constants.kt:4]: done!

File: app/src/main/java/com/allocate/ontime/MainActivity.kt [hunk 29]
```diff
@@ -41,13 +39,12 @@ class MainActivity : ComponentActivity() {
         const val LOCK_ACTIVITY_KEY = "com.allocate.ontime.MainActivity"
     }
 
-
-    @SuppressLint("CoroutineCreationDuringComposition")
     override fun onCreate(savedInstanceState: Bundle?) {
         super.onCreate(savedInstanceState)
         setContent {
-            OnTimeApp()
+            OnTimeApp(applicationContext)
```
Comment (mohit2988) [inline app/src/main/java/com/allocate/ontime/MainActivity.kt:45]: Why are we passing context? Can't we use `LocalContext.current` directly in composable?

File: app/src/main/java/com/allocate/ontime/business_logic/di/AppModule.kt [hunk 30]
```diff
@@ -35,22 +72,40 @@ object AppModule {
...
     }
 
-    @Singleton
     @Provides
-    fun provideRetrofitClient(): Retrofit {
+    @SuperAdminRetrofit
+    fun provideRetrofitClient2(@ApplicationContext context: Context): Retrofit {
+        val asApiUrl = SecureSharedPrefs(context).getData(
+            Constants.AS_API_URL,
+            "http://airstack.sublimitysoft.com/AirstackAPI/API/"
```
Comment (mohit2988) [inline app/src/main/java/com/allocate/ontime/business_logic/di/AppModule.kt:105]: Remove this from here and add it in some common place where we are keeping all the urls

File: app/src/main/java/com/allocate/ontime/business_logic/di/AppModule.kt [hunk 31]
```diff
@@ -22,8 +32,35 @@ import javax.inject.Singleton
...
+    @Provides
+    fun provideMainViewModelContext(
+        deviceInfoRepository: DeviceInfoRepository,
+        @ApplicationContext context: Context,
+    ) =
+        MainViewModel(deviceInfoRepository, context)
+
+    @Singleton
+    @Provides
+    fun provideDeviceUtils(
```
Comment (mohit2988) [inline app/src/main/java/com/allocate/ontime/business_logic/di/AppModule.kt:60]: Is It necessary to add this method in module class? as you already injected dependency with @Inject in the DeviceUtil class

File: app/src/main/java/com/allocate/ontime/business_logic/viewmodel/splash/SplashViewModel.kt (hunk 17 above)
[the remark this answers is not in the document]
Comment (mohit2988) [reply app/src/main/java/com/allocate/ontime/business_logic/viewmodel/splash/SplashViewModel.kt:89]: make it context private val and use it, no need to pass it as we are using it in same class

File: app/src/main/java/com/allocate/ontime/presentation_logic/navigation/OnTimeNavigation.kt [hunk 32]
```diff
@@ -20,7 +22,7 @@ import com.allocate.ontime.presentation_logic.screens.splash.SplashScreen
 
 @ExperimentalComposeUiApi
 @Composable
-fun OnTimeNavigation(viewModel: SuperAdminSettingViewModel) {
+fun OnTimeNavigation(context: Context, viewModel: SuperAdminSettingViewModel = hiltViewModel()) {
```
Comment (mohit2988) [inline app/src/main/java/com/allocate/ontime/presentation_logic/navigation/OnTimeNavigation.kt:25]: use this if you need context = LocalContext.current

File: app/src/main/java/com/allocate/ontime/business_logic/utils/DeviceUtils.kt [hunk 33]
```diff
@@ -0,0 +1,39 @@
...
+import android.net.wifi.WifiInfo
+import android.net.wifi.WifiManager
+import android.os.Build
+import android.telephony.TelephonyManager
+import java.util.Locale
+import javax.inject.Inject
+import javax.inject.Singleton
+
+@Singleton
+class DeviceUtils @Inject constructor(context: Context) {
```
Comment (mohit2988) [inline app/src/main/java/com/allocate/ontime/business_logic/utils/DeviceUtils.kt:14]: Rename : DeviceUtility

File: app/src/main/java/com/allocate/ontime/business_logic/di/AppModule.kt [hunk 34]
```diff
@@ -22,8 +32,35 @@ import javax.inject.Singleton
 object AppModule {
     // It provides the dependency of OnTimeRepository Class.
     @Provides
-    fun provideOnTimeRepository(api: OnTimeApi) =
-        OnTimeRepository(api)
+    fun provideDeviceInfoRepository(
```
Comment (mohit2988) [inline app/src/main/java/com/allocate/ontime/business_logic/di/AppModule.kt:35]: same for this https://github.com/sublimitysoftwares/OnTime_Android/pull/4/files#r1530113957

File: app/src/main/java/com/allocate/ontime/business_logic/di/AppModule.kt (hunk 30 above)
Comment (JitendraSaini1995) [reply app/src/main/java/com/allocate/ontime/business_logic/di/AppModule.kt:105]: done!

File: app/src/main/java/com/allocate/ontime/business_logic/di/AppModule.kt (hunk 31 above)
Comment (JitendraSaini1995) [reply app/src/main/java/com/allocate/ontime/business_logic/di/AppModule.kt:60]: done!

File: app/src/main/java/com/allocate/ontime/presentation_logic/navigation/OnTimeNavigation.kt (hunk 32 above)
Comment (JitendraSaini1995) [reply app/src/main/java/com/allocate/ontime/presentation_logic/navigation/OnTimeNavigation.kt:25]: done!

File: app/src/main/java/com/allocate/ontime/business_logic/utils/DeviceUtils.kt (hunk 33 above)
Comment (JitendraSaini1995) [reply app/src/main/java/com/allocate/ontime/business_logic/utils/DeviceUtils.kt:14]: done!

File: app/src/main/java/com/allocate/ontime/business_logic/di/AppModule.kt (hunk 34 above)
Comment (JitendraSaini1995) [reply app/src/main/java/com/allocate/ontime/business_logic/di/AppModule.kt:35]: done!

File: app/src/main/java/com/allocate/ontime/MainActivity.kt (hunk 29 above)
Comment (JitendraSaini1995) [reply app/src/main/java/com/allocate/ontime/MainActivity.kt:45]: done!

**Review by mohit2988 — APPROVED** [review]
Approved, Before merge test this code, if it's working then merge the code

**Pull request merged**